lib/services/local/saml_idp_service_provider.go

```diff
@@ -97,43 +103,59 @@ func (s *SAMLIdPServiceProviderService) CreateSAMLIdPServiceProvider(ctx context
 	if err := sp.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
-	value, err := services.MarshalSAMLIdPServiceProvider(sp)
-	if err != nil {
+
+	if err := s.ensureEntityDescriptorMatchesEntityID(sp); err != nil {
 		return trace.Wrap(err)
 	}
-	item := backend.Item{
-		Key:     backend.Key(samlIDPServiceProviderPrefix, sp.GetName()),
-		Value:   value,
-		Expires: sp.Expiry(),
-		ID:      sp.GetResourceID(),
-	}
-	_, err = s.Create(ctx, item)
-	if err != nil {
+
+	return trace.Wrap(backend.RunWhileLocked(ctx, s.Backend, samlIDPServiceProviderModifyLock, samlIDPServiceProviderModifyLockTTL, func(ctx context.Context) error {
+		if err := s.ensureEntityIDIsUnique(ctx, sp); err != nil {
+			return trace.Wrap(err)
+		}
+
+		value, err := services.MarshalSAMLIdPServiceProvider(sp)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		item := backend.Item{
+			Key:     backend.Key(samlIDPServiceProviderPrefix, sp.GetName()),
+			Value:   value,
+			Expires: sp.Expiry(),
+			ID:      sp.GetResourceID(),
+		}
```

Comment:
I'd prepare the `item` outside of the locked section, just as a matter of cleanliness; same applies below.